src/scanspec/core.py

```diff
@@ -104,72 +114,147 @@ def calculate(self) -> int:
-        cls.__annotations__ = {
-            **cls.__annotations__,
-            discriminator: Literal[cls.__name__],
-        }
-        setattr(cls, discriminator, field(default=cls.__name__, repr=False))
-
-    def __get_validators__(cls) -> Any:
-        yield cls.__validate__
-
-    def __validate__(cls, v: Any) -> Any:
-        # Lazily initialize model on first use because this
-        # needs to be done once, after all subclasses have been
-        # declared
-        if cls._model is None:
-            root = Union[tuple(cls._ref_classes)]  # type: ignore  # noqa
-            cls._model = create_model(
-                super_cls.__name__,
-                __root__=(root, Field(..., discriminator=discriminator)),
-                __config__=config,
-            )
-
-        try:
-            return cls._model(__root__=v).__root__
-        except ValidationError as e:
-            for (
-                error
-            ) in e.raw_errors:  # need in to remove redundant __root__ from error path
-                if (
-                    isinstance(error, ErrorWrapper)
-                    and error.loc_tuple()[0] == "__root__"
-                ):
-                    error._loc = error.loc_tuple()[1:]
-
-            raise e
-
-    # Inject magic methods into super_cls
-    for method in __init_subclass__, __get_validators__, __validate__:
-        setattr(super_cls, method.__name__, classmethod(method))  # type: ignore
-
-    return super_cls
+T = TypeVar("T", type, Callable)
+
+
+def deserialize_as(cls, obj):
+    return _tagged_unions[cls].type_adapter.validate_python(obj)
+
+
+def uses_tagged_union(cls_or_func: T) -> T:
+    """
+    Decorator that processes the type hints of a class or function to detect and
+    register any tagged unions. If a tagged union is detected in the type hints,
+    it registers the class or function as a referrer to that tagged union.
+
+    Args:
+        cls_or_func (T): The class or function to be processed for tagged unions.
+
+    Returns:
+        T: The original class or function, unmodified.
+
+    The function works by iterating over the type hints of the provided class or
+    function For each type hint, it checks if the type (or its origin,
+    in the case of generic types) is registered as a tagged union in the
+    global `_tagged_unions` dictionary. If a match is found, the
+    tagged union's `add_referrer` method is called to register the class
+    or function as a referrer.
+
+    Example:
+            class PointsRequest:
+            A request for generated scan points
+                spec: Spec
+        This will register `PointsRequest` and `Spec` with the corresponding tagged
+        union or
+            @uses_tagged_union
+            def valid(
+            spec: Spec = Body(..., examples=[_EXAMPLE_SPEC]),
+            ) -> ValidResponse | JSONResponse:
+            The decorator helps pydantic in understanding the core schema for the
+            function
```

Comment:
I'd nix this, it's implementation details. Docs are for how/when to use something, not what it does: how it does it should be allowed to change without the docstring needing to be updated.

```suggestion
```